Add map_from_entries Presto function

darrenfu opened the pull request:
## Summary
* Adding the `map_from_entries` Presto function to Velox

`map_from_entries(array(row(K, V))) -> map(K, V)`

> Returns a map created from the given array of entries.

For example:
```
SELECT map_from_entries(ARRAY[(1, 'x'), (2, 'y')]); -- {1 -> 'x', 2 -> 'y'}
SELECT map_from_entries(ARRAY[(1, 'x'), (2, null)]); -- {1 -> 'x', 2 -> null}
SELECT map_from_entries(ARRAY[(1, 'x'), (1, 'y')]); -- duplicate key error
SELECT map_from_entries(ARRAY[(null, 'x'), (1, 'y')]); -- map key cannot be null error
SELECT map_from_entries(ARRAY[cast(null as ROW(int, varchar)), (1, 'y')]); -- map entry cannot be null error
```

darrenfu commented:
> @darrenfu CI is failing. Would you take a look?

The CI error is a discrepancy between the two eval and evalSimplify paths:  
The latter will remove all the initial nulls (see https://github.com/facebookincubator/velox/pull/156), thus it will not throw an VeloxUserException if there is null in the input array. I believe this is unexpected because udf map_from_entries does not take a null map entry.

@pedroerp - since you are the author of https://github.com/facebookincubator/velox/pull/156, any advice to satisfy the fuzzer on both paths (while still throw a VeloxUserException for any nulls in the input array)?

mbasmanova commented on velox/docs/functions/map.rst:
```diff
@@ -26,6 +26,12 @@ Map Functions
 
     See also :func:`map_agg` for creating a map as an aggregation.
 
+.. function:: map_from_entries(array(row(K, V))) -> map(K, V)
```
Functions in the documentation should appear in alphabetical order. Hence, map_from_entries should come after map_entries. Please, update.

darrenfu replied:
> Functions in the documentation should appear in alphabetical order. Hence, map_from_entries should come after map_entries.

I think the function order in the documentation follows the [presto doc order](https://prestodb.io/docs/current/functions/map.html?highlight=map_zip_with#map_from_entries) where `map_from_entries` is ahead of `map_entries`. It doesn't follow alphabetical order. An anti example is: `map_entries` -> `map_concat` -> `map_filter`.

mbasmanova replied:
This looks like a bug in Presto's doc.

mbasmanova replied:
I'm not sure it makes sense to introduce this function in the wrong place. Velox and Presto are separate projects. We don't need to duplicate Presto bugs :-)

darrenfu replied:
Ok, I'm happy to adjust the function order in map.rst.

mbasmanova replied:
Thank you.

mbasmanova commented on velox/functions/prestosql/tests/MapFromEntriesTest.cpp:
```diff
@@ -0,0 +1,131 @@
+using namespace facebook::velox::functions::test;
+
+namespace {
+class MapFromEntriesTest : public FunctionBaseTest {
```
This function can be implemented in a zero-copy fashion as a vector function. We can reuse the input's offsets and sizes buffers as well as keys and values vectors to construct the result map. Vector function implementation will also allow us to support all types, not just primitive types for input. See velox/functions/prestosql/Map.cpp for an example code for checking for duplicate keys.

darrenfu replied:
Thanks for looking at my updated diff, @mbasmanova .

I understand your concern about lack of zero copy optimization for the generic-typed input. More general speaking, would you help me understand better the team criteria when a velox developer should use simple function api vs vector function api? I raised this question because you also suggest to reimplement `shuffle(array)` via vector function api.

Is supporting generic typed input/output + zero copy optimization the ONLY criteria?

mbasmanova replied:
We recommend to use simple function API whenever possible. However, if simple function API cannot be used to implement all the necessary functionality or if vector function can be significantly faster, then vector function API should be used.

mbasmanova reviewed (changes requested):
@darrenfu I think this function needs to be implemented as a vector function. See comment below for details. Also, to test this new function more thoroughly, please, run the fuzzer. First, test function in isolation, then in combination with other functions.

```
velox_expression_fuzzer_test --logtostderr=1 --velox_fuzzer_enable_complex_types --only map_from_entries --duration_sec 60

velox_expression_fuzzer_test --logtostderr=1 --velox_fuzzer_enable_complex_types --enable_variadic_signatures --duration_sec 600
```

mbasmanova commented:
@darrenfu CI is red. Would you take a look? Make sure to rebase onto latest main to pick all the recent fixes.

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,105 @@
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    // array(row(K,V)) -> map(K,V)
+    return {exec::FunctionSignatureBuilder()
+                .knownTypeVariable("K")
```
Presto allows an array(unknown) as input:

```
presto:di> select map_from_entries(array[]);
 _col0
-------
 {}

presto:di> select typeof(map_from_entries(array[]));
         _col0
-----------------------
 map(unknown, unknown)
(1 row)
```

darrenfu replied:
Ok. Not super familiar with unknown type signature handling, is there any example to follow?

mbasmanova replied:
Here are some examples:

velox/functions/prestosql/ArrayConstructor.cpp

```
        // () -> array(unknown)
        exec::FunctionSignatureBuilder().returnType("array(unknown)").build(),
        // T... -> array(T)
        exec::FunctionSignatureBuilder()
            .typeVariable("T")
            .returnType("array(T)")
            .argumentType("T")
            .variableArity()
            .build(),
```

velox/functions/prestosql/InPredicate.cpp

```
      signatures.emplace_back(exec::FunctionSignatureBuilder()
                                  .returnType("boolean")
                                  .argumentType(type)
                                  .argumentType("array(unknown)")
                                  .build());
```

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,105 @@
+    auto& arg = args[0];
+    MapVectorPtr mapVector;
+
+    auto inputArray = arg->as<ArrayVector>();
```
Input for a single-arg function can be either flat or constant. Hence, the function needs to handle both flat and constant encodings.

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,105 @@
+
+    auto inputArray = arg->as<ArrayVector>();
+    auto rowVector = inputArray->elements();
+    auto rowKeys = rowVector->as<RowVector>()->childAt(0);
```
The elements vector may have encodings. It doesn't have to be flat, hence, the function needs to handle all encodings.

darrenfu replied:
Do you mean we also want to handle both flat and constant encoding for the elements vector (RowVector)? I thought handling the top level (single-arg of array vector) should suffice.

mbasmanova replied:
The top-level argument is guaranteed to be either flat or constant, but nested vectors can have any encoding (flat, constant or dictionary).

darrenfu replied:
I may need a quick chat to help understand how to handle such nested vector encodings in this case.

Since this udf takes one argument: `array(row(K, V))`, the encodings will be something like below, correct? 

| Level  | variable | vector encoding|
| ------------- | ------------- | ------------- |
| top level  | args[0] | constant / flat < ComplexType >  |
| second level  | *topArrayVectorPtr->elements() | constant / flat / dict < ComplexType >  |

Why aren't the vector encodings just `array` and `row`? Appreciate if there is any document to help me understand the mappings between vector types and encodings.

I can find the [example code](https://github.com/darrenfu/velox/blob/dd955aaddff0601a23c56d1ad62d0917b3aa2f94/velox/functions/prestosql/MapKeysAndValues.cpp#L35-L47) for the top level constant/flat handling. But for the second level (e.g. row vector), there are no similar examples. Is it something like below? How shall we handle dict vector encoding? 

```cpp
auto inputArray = decodedArg->base()->as<ArrayVector>();
auto rowVector = inputArray->elements();
switch (rowVector->encoding()) {
  case CONSTANT: {
    // repeat the same logic where top level vector is a constant vector
    auto* constantRow = arg->as<ConstantVector<ComplexType>>();
    const auto& flatRow = constantRow->valueVector();
    ...
    break;
  }
  case FLAT: {
    auto actualRowVector = rowVector->as<RowVector>();
    ...
    break;
  }
  case DICTIONARY:
    // How to handle this encoding? 
    break;
}
```

mbasmanova replied:
@darrenfu Darren, feel free to ping me offline or book some VC time.

Encodings and types are independent: https://facebookincubator.github.io/velox/develop/vectors.html 

A canonical way to handle encodings is to use DecodedVector. In Vector functions, we often introduce fast paths for common use cases.

darrenfu replied:
Thanks @mbasmanova ! I booked 30 min on your calendar. Let's sync to clear some of my roadblocks to accelerate velox development then!

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,105 @@
+    auto rowKeys = rowVector->as<RowVector>()->childAt(0);
+    auto rowValues = rowVector->as<RowVector>()->childAt(1);
+
+    DecodedVector rowVectorDecoded;
```
Use DecodedArgs (or LocalDecodedVector).

mbasmanova replied:
it is better to use LocalDecodedVector (it allows to avoid extra memory allocations). Even better - DecodedArgs (less core, more readable).

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,105 @@
+    auto offsets = inputArray->rawOffsets();
+    auto sizes = inputArray->rawSizes();
+
+    rows.applyToSelected([&](vector_size_t row) {
```
To support TRY expression, use context.applytoSelectedNoThrow: https://facebookincubator.github.io/velox/develop/scalar-functions.html#try-expression-support

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,105 @@
+      auto offset = offsets[row];
+      auto size = sizes[row];
+      for (vector_size_t i = 1; i < size; i++) {
+        if (rowKeys->equalValueAt(rowKeys.get(), offset + i, offset + i - 1)) {
```
The elements of the array may not be sorted, hence, this check is incorrect. See velox/functions/prestosql/Map.cpp for an example of how to check for duplicate keys in a map after constructing the MapVector.

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,105 @@
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    auto& arg = args[0];
+    MapVectorPtr mapVector;
```
move this variable to where it is first used

mbasmanova reviewed (commented):
@darrenfu Some comments about the implementation. I expect the fuzzer test to fail for this function because it doesn't handle encodings properly. Curious to know if it does.

mbasmanova reviewed (changes requested):
@darrenfu Darren, thank you for iterating on this PR. Working with complex types and encodings is indeed a bit tricky. Here are some suggestions.

darrenfu commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,147 @@
+      localResult =
+          BaseVector::wrapInConstant(rows.size(), flatIndex, localResult);
+    } else {
+      exec::DecodedArgs decodedArgs(rows, args, context);
```
Thank you!!! Your sketch is super helpful for me to decode such a nested complex vector.  

FYI: I know the gist of different simple vector layouts when I was in Saber, but I am not super familiar with the surrounding helper / wrapper classes and the usage scenarios, like SelectivityVector, complex vector layouts (this is not contributed from Saber), DecodedVector (I wrote our own vector SerDe classes when I was in Saber). So excited to witness so many new velox features since then, and happy to refresh my arsenal.

Also thanks for your quick workaround for bug #3424 and #3533, I amended a comment - it actually affects a few map functions.

darrenfu replied:
> I modified the Fuzzer to workaround https://github.com/facebookincubator/velox/issues/3533 and also modified DecodedVector to workaround a bug that @kagamiori Wei is looking into (https://github.com/facebookincubator/velox/pull/3424)

Since Wei is working on the fuzzer bugs, I will leave the fuzzer code unchanged with CI failures. Once these two bugs are fixed, I will rebase from main and land this PR. Does it sound good?

mbasmanova replied:
> Since Wei is working on the fuzzer bugs, I will leave the fuzzer code unchanged with CI failures. Once these two bugs are fixed, I will rebase from main and land this PR. Does it sound good?

Indeed, this PR has a dependency on fixing Fuzzer and DecodedVector. Just to clarify, after rebasing the PR, you'll need to re-run Fuzzer to make sure there are not failures and the PR will need to be reviewed before it is ready for landing.

darrenfu replied:
@mbasmanova - regarding the fuzzer failure, I did some debugging and found that the exception is thrown at this line:
```c++
    VectorPtr wrappedKeys;
    VectorPtr wrappedValues;
    if (decodedRow->isIdentityMapping()) {
      wrappedKeys = rowVector->childAt(0);
      wrappedValues = rowVector->childAt(1);
    } else {
      wrappedKeys = decodedRow->wrap(
          rowVector->childAt(0), *inputRowVector, inputRowVector->size()); // <---- THROW EXCEPTION: 
          // "Dictionary index must be less than base vector's size. Index:..." 
          // mostly reproduced at the simplified path + the duplicate-key case
      wrappedValues = decodedRow->wrap(
          rowVector->childAt(1), *inputRowVector, inputRowVector->size());
    }
```
Question:
Why do you call `decodedRow->wrap()` to wrap the keys and values vectors for non-flat encodings in your [sketch code](https://github.com/facebookincubator/velox/compare/main...mbasmanova:velox-1:map_from_entries_simpleudf?expand=1)? Maybe I'm wrong, but I feel we should compute the total size, new indices, and call `BaseVector::wrapInDictionary()` to wrap the keys and values. Refer to [map()'s implementation](https://github.com/facebookincubator/velox/blob/22d9efb0645801aa2f193740827829c088210884/velox/functions/prestosql/Map.cpp#L163-L173). What do you think?

I noticed that this edge case usually happens when keys is a nested dictionary encoding or even more complex encodings.

darrenfu replied:
Found the root cause, see https://github.com/facebookincubator/velox/issues/3687

darrenfu commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,147 @@
+    return mapVector;
+  }
+
+  static void checkForDuplicateKeys(
```
Thanks for extracting this common utility function for map udfs!

darrenfu commented:
The current CI failure is suspicious to fuzzer's lack of support on nested complex types, aka. ARRAY(ROW<>).
See more discussion: https://github.com/facebookincubator/velox/issues/3388#issuecomment-1355383398

cc @bikramSingh91 for awareness.

bikramSingh91 commented:
> The current CI failure is suspicious to fuzzer's lack of support on nested complex types, aka. ARRAY(ROW<>). See more discussion: [#3388 (comment)](https://github.com/facebookincubator/velox/issues/3388#issuecomment-1355383398)
> 
> cc @bikramSingh91 for awareness.

@darrenfu #3540 recently landed which now also picks return types for top level expressions from functions that have templated return types. therefore, it is now possible to pick zip with return type array(row(T,K,..))  and randomly bind the templated types to concrete types.

kagamiori commented:
> > The current CI failure is suspicious to fuzzer's lack of support on nested complex types, aka. ARRAY(ROW<>). See more discussion: [#3388 (comment)](https://github.com/facebookincubator/velox/issues/3388#issuecomment-1355383398)
> > cc @bikramSingh91 for awareness.
> 
> @darrenfu #3540 recently landed which now also picks return types for top level expressions from functions that have templated return types. therefore, it is now possible to pick zip with return type array(row(T,K,..)) and randomly bind the templated types to concrete types.

Sorry I haven't landed #3540 yet. It was blocked by #3564. I expect to land #3540 soon.

bikramSingh91 commented:
> > > The current CI failure is suspicious to fuzzer's lack of support on nested complex types, aka. ARRAY(ROW<>). See more discussion: [#3388 (comment)](https://github.com/facebookincubator/velox/issues/3388#issuecomment-1355383398)
> > > cc @bikramSingh91 for awareness.
> > 
> > 
> > @darrenfu #3540 recently landed which now also picks return types for top level expressions from functions that have templated return types. therefore, it is now possible to pick zip with return type array(row(T,K,..)) and randomly bind the templated types to concrete types.
> 
> Sorry I haven't landed #3540 yet. It was blocked by #3564. I expect to land #3540 soon.

Ah, Interesting, let me dive into this tomorrow and try to figure out what is happening then.

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,141 @@
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/expression/VectorReaders.h"
```
Is this header needed?

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,141 @@
+      const auto& flatArray = constantArray->valueVector();
+      const auto flatIndex = constantArray->index();
+
+      exec::LocalSelectivityVector singleRow(context, flatIndex + 1);
```
same comments as for shuffle function

mbasmanova commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,141 @@
+    auto rowVector = decodedRow.get()->base()->as<RowVector>();
+    auto rowKeyVector = rowVector->childAt(0);
+
+    // validate all map entries and map keys are not null
```
Comments should be full sentences, i.e. start with a capital letter and end with a period. Please, check and update throughout.

mbasmanova commented:
@darrenfu CircleCI is failing and the failure seems related to the new function. Would you take a look?

bikramSingh91 commented:
> @darrenfu CircleCI is failing and the failure seems related to the new function. Would you take a look?

I've detailed the root cause and fix of the issue here, please take a look.
https://github.com/facebookincubator/velox/issues/3581#issuecomment-1363474314

mbasmanova commented on velox/functions/prestosql/tests/MapFromEntriesTest.cpp:
```diff
@@ -0,0 +1,202 @@
+  evaluateExprAndAssert(expected, "map_from_entries(C0)", {input});
+}
+
+TEST_F(MapFromEntriesTest, intKeyAndVarcharValue) {
```
I'm seeing multiple tests cases that differ only in the types of keys and values. Since the implementation of the function is agnostic to these, I think one test case is sufficient.

mbasmanova commented on velox/functions/prestosql/tests/MapFromEntriesTest.cpp:
```diff
@@ -0,0 +1,202 @@
+  auto input = makeArrayOfRowVector(rowType, data);
+  auto expected = makeMapVector<int32_t, StringView>(
+      {{{1, "red"_sv}, {2, "blue"_sv}, {3, "green"_sv}}});
+  evaluateExprAndAssert(expected, "map_from_entries(C0)", {input});
```
I wonder if it would be cleaner to refactor this as

```
verifyMapFromEntries(input, expected);
```

mbasmanova commented on velox/functions/prestosql/tests/MapFromEntriesTest.cpp:
```diff
@@ -0,0 +1,202 @@
+  VELOX_ASSERT_THROW(
+      evaluateExpr("map_from_entries(C0)", {input}),
+      "map entry cannot be null");
+  EXPECT_NO_THROW(evaluateExpr("try(map_from_entries(C0))", {input}));
```
It would be nice to verify the results of the try as well? E.g. generate input where some rows generate errors and some don't and verify that rows with error return null and other rows return correct results.

darrenfu replied:
That's a good idea. Let me amend it.

mbasmanova commented on velox/functions/prestosql/tests/MapFromEntriesTest.cpp:
```diff
@@ -0,0 +1,202 @@
+  evaluateExprAndAssert(expected, "map_from_entries(C0)", {input});
+}
+
+TEST_F(MapFromEntriesTest, moreThanTwoColumnsInRow) {
```
This test case is not necessary.

darrenfu replied:
Why do you think this edge case is unnecessary? Is it guarded by the function `signature()` definition?

mbasmanova reviewed (commented):
@darrenfu Darren, thank you for iterating on the PR.

The function implementation looks good to me. Some comments about the tests.

mbasmanova commented on velox/docs/functions/json.rst:
```diff
@@ -89,18 +89,6 @@ special characters unescaped and double quotes removed.
 JSON Functions
 --------------
 
-.. function:: json_extract_scalar(json, json_path) -> varchar
```
Thank you for the cleanup. Since this is an unrelated change, would you submit it as a separate PR?

mbasmanova commented:
@darrenfu Darren, CI is failing on this PR. Would you take a look?

darrenfu commented:
Update: I will resume the work on this PR.

pranjalssh commented on velox/vector/tests/utils/VectorTestBase.h:
```diff
@@ -244,9 +244,7 @@ class VectorTestBase {
 
   // Create an ArrayVector<ROW> from nested std::vectors of variants.
   // Example:
-  //   auto arrayVector = makeArrayOfRowVector(
```
Why is this removed?

kagamiori commented on velox/functions/prestosql/tests/MapFromEntriesTest.cpp:
```diff
@@ -0,0 +1,161 @@
+namespace {
+class MapFromEntriesTest : public FunctionBaseTest {
+ protected:
+  void verifyMapFromEntriesWithTry(
+      const VectorPtr& input,
+      const VectorPtr& expected) {
+    verifyMapFromEntries(input, expected, true);
+  }
```
nit: No need to add a function with a single-line body. You can just call verifyMapFromEntries(input, expected, true) in the caller. Ditto for evaluateExpr.

darrenfu replied:
I agree with calling verifyMapFromEntries() instead. 
For evaluateExpr(), I think it's necessary to encapsulate the std::vector<VectorPtr> to RowVectorPtr conversion in a utility function, which is consistent with other utility functions in this class. See all the call sites of makeRowVector().

kagamiori commented on velox/functions/prestosql/tests/MapFromEntriesTest.cpp:
```diff
@@ -0,0 +1,161 @@
+      evaluateExpr("map_from_entries(C0)", {input}),
+      "map entry cannot be null");
+  auto expected =
+      makeNullableMapVector<int32_t, int32_t>({std::nullopt, O({{1, 11}})});
```
Does it work to replace `O()` with an additional pair of `{}` and avoid the use of O() for readability?

darrenfu replied:
I don't think we can initialize a std::optional object by calling something like `{n}`, which is an initializer list, not a single value. 
There are also some existing code pointer I borrowed from, e.g. [this](https://github.com/facebookincubator/velox/blob/main/velox/vector/tests/VectorMakerTest.cpp#L347-L347).

kagamiori replied:
You can try it out, because I know in some tests we use `{n}` to initialize std::optional. If it doesn't work here, you can keep O().

kagamiori commented on velox/functions/prestosql/tests/MapFromEntriesTest.cpp:
```diff
@@ -0,0 +1,161 @@
+          {{{"r"_sv, 11}, {"g"_sv, 22}, {"b"_sv, 33}}})};
+
+  evaluateExprWithConstantArray(expected, "map_from_entries(C0)", input);
+}
```
Could you add a unit test where the row vector inside the array vector and the keys vector are not flat vectors?

kagamiori replied:
Hi @darrenfu, It's different from evaluateExprWithConstantArray() which only make the top-level array vector in constant encoding. What I was suggesting is that since applyFlat handles different encoding of the element vector (i.e., the row vector) of the input array vector, it'd be better to have a test for this too.

darrenfu replied:
@kagamiori - I've added two new unit test cases to cover your case (`dictionaryEncodedElementsInFlat`) and the case  (`outputSizeIsBoundBySelectedRows`) where I fixed a bug (similar to #4455 ).

kagamiori reviewed (commented):
Hi Darren, Thank you for adding this function! It looks good to me overall. I left a few comments. Please also run fuzzer for 30 minutes to ensure no failures.

darrenfu commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,139 @@
+      exec::EvalCtx& context) const {
+    auto& inputRowVector = inputArray->elements();
+    exec::LocalDecodedVector decodedRow(context);
+    decodedRow.get()->decode(*inputRowVector);
```
I can remove .get() for decodedRow->base().
I still need .get() for decodedRow.get()->decode() because its `vector_` hasn't been initialized yet before decoding.

darrenfu commented:
cc @kgpai for the benchmark result upload failure (503 error). I hope it's a transient error.

kgpai commented:
@darrenfu It was transient, though we are still working on that job and need to update the keys that would allow us to upload the results.

darrenfu commented:
> Hi Darren, Thank you for adding this function! It looks good to me overall. I left a few comments. Please also run fuzzer for 30 minutes to ensure no failures.

Thanks for mentioning that test approach. Yes, I just did locally. Nothing interesting is thrown.

darrenfu commented:
> @darrenfu Darren, thank you for iterating on the PR.
> 
> The function implementation looks good to me. Some comments about the tests.

Addressed all the comments for the unit tests and fuzz test. I believe the PR is in a ready-to-merge state. Let me know if there is anything else to address.

xiaoxmeng commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,136 @@
+        for (auto i = 0; i < size; ++i) {
+          bool isMapEntryNull = decodedRow->isNullAt(offset + i);
+          VELOX_USER_CHECK(!isMapEntryNull, "map entry cannot be null");
+          bool isMapKeyNull =
```
```
nit: const bool isMapKeyNull =
```

xiaoxmeng commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,136 @@
+        auto size = inputArray->sizeAt(row);
+        auto offset = inputArray->offsetAt(row);
+        for (auto i = 0; i < size; ++i) {
+          bool isMapEntryNull = decodedRow->isNullAt(offset + i);
```
```
nit: const bool isMapEntryNull
```

xiaoxmeng commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,136 @@
+      wrappedKeys = rowVector->childAt(0);
+      wrappedValues = rowVector->childAt(1);
+    } else {
+      wrappedKeys = decodedRow->wrap(
```
Where we handle the duplicate keys? Thanks!

darrenfu replied:
It is handled via calling `checkDuplicateKeys()`

darrenfu commented on velox/functions/prestosql/MapFromEntries.cpp:
```diff
@@ -0,0 +1,136 @@
+      context.applyToSelectedNoThrow(rows, [&](vector_size_t row) {
+        auto size = inputArray->sizeAt(row);
+        auto offset = inputArray->offsetAt(row);
+        for (auto i = 0; i < size; ++i) {
```
Can we do that by just checking size == 0? The purpose is to validate all map entries and map keys are not null here.
